### PR TITLE
Added the NCCM datamodule

### PR DESCRIPTION
**Pull request by shreyakannan1205**
The purpose of this PR is to added the datamodule for the Northeastern China Crop Map dataset.

### CODE REVIEW

File: torchgeo/datamodules/nccm.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+            },
+        )
+
+    def setup(self, stage: str) -> None:
+        """Set up datasets.
+
+        Args:
+            stage: Either 'fit', 'validate', 'test', or 'predict'.
+        """
+        dataset = NCCM(**self.kwargs)
```
Comment (adamjstewart) [inline torchgeo/datamodules/nccm.py:69]: NCCM only has masks, you'll need to pair it with a Sentinel2 dataset in order to train.

**Review by adamjstewart — COMMENTED** [review]
Needs import in `__init__.py` and tests. Once you have both images and masks, tests are easy, just add a config file in `tests/conf/nccm.yaml` and add a new line to `tests/trainers/test_segmentation.py`.

File: tests/conf/nccmsentinel2.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+model:
+  class_path: SemanticSegmentationTask
+  init_args:
+    loss: "ce"
+    model: "deeplabv3+"
+    backbone: "resnet34"
+    in_channels: 4
+    num_classes: 14
+    num_filters: 1
+    ignore_index: null
```
Comment (yichiac) [inline tests/conf/nccmsentinel2.yaml:10]:
```suggestion
    num_classes: 4
    num_filters: 1
    ignore_index: 4
```
The nodata class is 4, which should be ignored during training. The actually classes are class 0-3.

**Review by yichiac — CHANGES_REQUESTED** [review]

File: tests/conf/nccmsentinel2.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    ignore_index: 4
+data:
+  class_path: NCCMSentinel2DataModule
+  init_args:
+    batch_size: 2
+    patch_size: 32
+  dict_kwargs:
+    nccm_paths: "tests/data/nccm"
+    sentinel2_paths: "tests/data/sentinel2"
+
```
Comment (adamjstewart) [inline tests/conf/nccmsentinel2.yaml:19]: You can delete this empty line

File: tests/data/nccm/data.py
Comment (adamjstewart) [inline tests/data/nccm/data.py:1]: You'll also need to re-run this file and update the files for these changes to take affect.

File: torchgeo/datamodules/nccm.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+                :class:`~torchgeo.datasets.Sentinel2`
+                (prefix keys with ``sentinel2_``).
+        """
+        self.nccm_kwargs = {}
+        self.sentinel2_kwargs = {}
+        for key, val in kwargs.items():
+            if key.startswith("nccm_"):
+                self.nccm_kwargs[key[5:]] = val
+            elif key.startswith("sentinel2_"):
+                self.sentinel2_kwargs[key[10:]] = val
```
Comment (adamjstewart) [inline torchgeo/datamodules/nccm.py:49]: If this starts to become a common kwarg parsing approach, we can add a utility function to avoid code duplication. But right now it's only used in 2 places.

File: torchgeo/datamodules/nccm.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        Args:
+            stage: Either 'fit', 'validate', 'test', or 'predict'.
+        """
+        self.sentinel2 = Sentinel2(**self.sentinel2_kwargs)
+        self.nccm = NCCM(**self.nccm_kwargs)
+        self.dataset = self.sentinel2 & self.nccm
+
+        roi = self.dataset.bounds
+        midx = roi.minx + (roi.maxx - roi.minx) / 2
+        midy = roi.miny + (roi.maxy - roi.miny) / 2
```
Comment (adamjstewart) [inline torchgeo/datamodules/nccm.py:71]: We should probably use one of the more advanced [splitting functions](https://torchgeo.readthedocs.io/en/stable/api/datasets.html#splitting-functions) now that they exist. @yichiac how do you want to perform a train-val-test split for this dataset for the paper?

Comment (adamjstewart) [reply torchgeo/datamodules/nccm.py:71]: It wouldn't change how efficient training is, the code would just be cleaner and you could guarantee a perfect 80-10-10 split instead of an approximate split.

File: tests/conf/nccmsentinel2.yaml [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    num_filters: 1
+    ignore_index: 4
+data:
+  class_path: NCCMSentinel2DataModule
+  init_args:
+    batch_size: 2
+    patch_size: 32
+  dict_kwargs:
+    nccm_paths: "tests/data/nccm"
+    sentinel2_paths: "tests/data/sentinel2"
```
Comment (adamjstewart) [inline tests/conf/nccmsentinel2.yaml:18]: Can you add an EOL char before the EOF char? It should be possible to set your editor to do this automatically.

File: torchgeo/datamodules/nccm.py (hunk 5 above)
Comment (adamjstewart) [reply torchgeo/datamodules/nccm.py:71]: Seems like this is the only remaining change required to merge this PR...

**Comment by shreyakannan1205** [comment]
I fixed the mapping issue, now the main issue that is remaining is the minx and maxx values for test_sampler. only for test_sampler, the minx value is +ve and maxx value is -ve. I am not sure on how to fix this. @adamjstewart @yichiac

File: .DS_Store
Comment (adamjstewart) [inline .DS_Store:1]: Please remove

File: tests/.DS_Store
Comment (adamjstewart) [inline tests/.DS_Store:1]: Please remove

File: torchgeo/.DS_Store
Comment (adamjstewart) [inline torchgeo/.DS_Store:1]: Please remove

File: torchgeo/trainers/segmentation.py [hunk 7]
```diff
@@ -252,8 +253,14 @@ def validation_step(
         ):
             datamodule = self.trainer.datamodule
             batch["prediction"] = y_hat.argmax(dim=1)
+
+            output, indices = torch.unique(
+                batch["prediction"], sorted=True, return_inverse=True
```
Comment (adamjstewart) [inline torchgeo/trainers/segmentation.py:258]: Why is this needed?

File: torchgeo/samplers/batch.py [hunk 8]
```diff
@@ -107,10 +107,12 @@ def __init__(
 
         self.batch_size = batch_size
         self.length = 0
+
```
Comment (adamjstewart) [inline torchgeo/samplers/batch.py:110]: Doesn't seem related to this PR

File: torchgeo/datamodules/l7irish.py [hunk 9]
```diff
@@ -68,9 +68,11 @@ def setup(self, stage: str) -> None:
...
         dataset = L7Irish(**self.kwargs)
         generator = torch.Generator().manual_seed(0)
-        (self.train_dataset, self.val_dataset, self.test_dataset) = (
-            random_bbox_assignment(dataset, [0.6, 0.2, 0.2], generator)
-        )
+        (
+            self.train_dataset,
+            self.val_dataset,
+            self.test_dataset,
+        ) = random_bbox_assignment(dataset, [0.6, 0.2, 0.2], generator)
```
Comment (adamjstewart) [inline torchgeo/datamodules/l7irish.py:75]: Make sure you're using the latest version of black, the formatting has changed in the 2024 version.

**Review by adamjstewart — CHANGES_REQUESTED** [review]

**Comment by adamjstewart** [comment]
> I fixed the mapping issue, now the main issue that is remaining is the minx and maxx values for test_sampler. only for test_sampler, the minx value is +ve and maxx value is -ve. I am not sure on how to fix this. @adamjstewart @yichiac need a bit of help to fix this! thanks

This is a red herring, you can ignore it. How big is your image size and how big is your patch size? I'm guessing that if you compute the number of patches that your image can be divided into, then 10% of that number rounds down to 0. So if your patch size is 16, try an image size of 64.

**Comment by shreyakannan1205** [comment]
> > I fixed the mapping issue, now the main issue that is remaining is the minx and maxx values for test_sampler. only for test_sampler, the minx value is +ve and maxx value is -ve. I am not sure on how to fix this. @adamjstewart @yichiac need a bit of help to fix this! thanks
> 
> This is a red herring, you can ignore it. How big is your image size and how big is your patch size? I'm guessing that if you compute the number of patches that your image can be divided into, then 10% of that number rounds down to 0. So if your patch size is 16, try an image size of 64.

**Comment by shreyakannan1205** [comment]
I changed the image size to 64 in data.py and uploaded the new data. The patch size is 16. The tests are still failing, so I am not sure on how to fix this. 
@adamjstewart @yichiac

File: torchgeo/datamodules/nccm.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+from ..datasets import NCCM, Sentinel2, random_bbox_assignment
+from ..samplers import GridGeoSampler, RandomBatchGeoSampler
+from ..transforms import AugmentationSequential
+from .geo import GeoDataModule
+
+
+class NCCMSentinel2DataModule(GeoDataModule):
+    """LightningDataModule implementation for the NCCM and Sentinel2 datasets.
+
+    Uses the train/val/test splits from the dataset.
```
Comment (adamjstewart) [inline torchgeo/datamodules/nccm.py:21]: Add:
```rst
.. versionadded:: 0.6
```

File: torchgeo/datamodules/nccm.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        """Run NCCM plot method.
+
+        Args:
+            *args: Arguments passed to plot method.
+            **kwargs: Keyword arguments passed to plot method.
+
+        Returns:
+            A matplotlib Figure with the image, ground truth, and predictions.
+
+        .. versionadded:: 0.4
```
Comment (adamjstewart) [inline torchgeo/datamodules/nccm.py:100]: Remove this, this data module did not exist in 0.4 and has always had a plot method

File: torchgeo/datamodules/nccm.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        if stage in ["fit"]:
+            self.train_batch_sampler = RandomBatchGeoSampler(
+                self.train_dataset, self.patch_size, self.batch_size, self.length
+            )
+
+        if stage in ["fit", "validate"]:
+            self.val_sampler = GridGeoSampler(
+                self.val_dataset, self.patch_size, self.patch_size
+            )
+        print("passed")
```
Comment (adamjstewart) [inline torchgeo/datamodules/nccm.py:84]: Remove print statements

File: torchgeo/datamodules/nccm.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+            if key.startswith("nccm_"):
+                self.nccm_kwargs[key[5:]] = val
+            elif key.startswith("sentinel2_"):
+                self.sentinel2_kwargs[key[10:]] = val
+
+        super().__init__(
+            NCCM, batch_size, patch_size, length, num_workers, **self.nccm_kwargs
+        )
+
+        self.aug = AugmentationSequential(
```
Comment (adamjstewart) [inline torchgeo/datamodules/nccm.py:56]: This doesn't do anything that the default augmentation doesn't do. @yichiac which augmentations do you plan to use?

File: torchgeo/samplers/batch.py [hunk 14]
```diff
@@ -111,6 +111,7 @@ def __init__(
         areas = []
         for hit in self.index.intersection(tuple(self.roi), objects=True):
             bounds = BoundingBox(*hit.bounds)
+
```
Comment (adamjstewart) [inline torchgeo/samplers/batch.py:114]: Unrelated change

File: torchgeo/trainers/segmentation.py [hunk 15]
```diff
@@ -252,8 +252,10 @@ def validation_step(
         ):
             datamodule = self.trainer.datamodule
             batch["prediction"] = y_hat.argmax(dim=1)
+
```
Comment (adamjstewart) [inline torchgeo/trainers/segmentation.py:255]: Unrelated change

File: torchgeo/datamodules/nccm.py (hunk 13 above)
Comment (yichiac) [reply torchgeo/datamodules/nccm.py:56]: Can we do all augmentations similar to `l7irish`?
```
self.train_aug = AugmentationSequential(
            K.Normalize(mean=self.mean, std=self.std),
            K.RandomResizedCrop(_to_tuple(self.patch_size), scale=(0.6, 1.0)),
            K.RandomVerticalFlip(p=0.5),
            K.RandomHorizontalFlip(p=0.5),
            data_keys=["image", "mask"],
            extra_args={
                DataKey.MASK: {"resample": Resample.NEAREST, "align_corners": None}
            },
        )
```

**Comment by adamjstewart** [comment]
> I changed the image size to 64 in data.py and uploaded the new data. The patch size is 16. The tests are still failing, so I am not sure on how to fix this.

The resulting IntersectionDataset contains 3 bbox entries. You're using `random_bbox_assignment` with an 80-10-10 split. 10% of 3 is 0, so the val and test datasets are empty. This is the cause of the error.

@yichiac how do you want to create the train-val-test split? I explained all the splitting options on GitHub. It really depends on what you're trying to study, they're all valid options.
